File: src/lib/controllers/Project.ts
import { invoke } from "@tauri-apps/api";
import { reloadRepository } from "./Repository";
import type { TreeItem } from "$lib/components/structs/Tree";
import type { Project, ProjectManifest } from "$lib/components/structs/Project";

export async function createProject(manifest: ProjectManifest, parent: TreeItem) {
    return invoke('create_project', {path: parent.path, man: manifest})
        .then((prj) => {
            reloadRepository();
            return prj as Project;
        })
}

export async function readProject(project: TreeItem) {
    return invoke('read_project', {path: project.path})
        .then((prj) => {
            return prj as Project;
        })
}

export async function updateProject(project: TreeItem, man: ProjectManifest) {
    return invoke('update_project', {path: project.path, man: man})
}

export async function deleteProject(project: TreeItem) {
    return invoke('delete_project', {path: project.path})
}

File: src-tauri/src/handlers/project.rs
use std::{path::PathBuf, sync::Mutex};
use git2::Repository;
use tauri::{command, State}; 

use crate::core::{error::OpenDoorsError, project::{Project, ProjectManifest}};

#[command] 
pub fn create_project(state: State<'_, Mutex<Option<Repository>>>, path: &str, man: ProjectManifest) -> Result<Project, OpenDoorsError> {
	let repo = state.lock().unwrap();
	Ok(Project::create(&repo, &path, &man)?)
}

#[command]
pub fn read_project(path: PathBuf) -> Result<Project, OpenDoorsError> {
	Ok(Project::read(&path)?)
}

#[command]
pub fn update_project(state: State<'_, Mutex<Option<Repository>>>, path: PathBuf, man: ProjectManifest) -> Result<Project, OpenDoorsError> {
	let repo = state.lock().unwrap();
	Project::update_manifest(&repo, &path, &man)?;
	Ok(Project::read(&path)?)
}

#[command]
pub fn delete_project(state: State<'_, Mutex<Option<Repository>>>, path: PathBuf) -> Result<(), OpenDoorsError> {
	let repo = state.lock().unwrap();
	Ok(Project::delete(&repo, &path)?)
}
